tools/wasm-sourcemap.py

```diff
@@ -142,173 +130,74 @@ def get_code_section_offset(wasm):
+  sources = map['sources']
+  sources_content = []
+  for source_id in range(len(sources)):
+    file_name = sources[source_id]
     source_name = prefixes.sources.resolve(file_name)
-    if source_name not in sources_map:
-      source_id = len(sources)
-      sources_map[source_name] = source_id
-      sources.append(source_name)
-      if collect_sources:
-        load_name = prefixes.load.resolve(file_name)
-        try:
-          with open(load_name, 'r') as infile:
-            source_content = infile.read()
-          sources_content.append(source_content)
-        except:
-          print('Failed to read source: %s' % load_name)
-          sources_content.append(None)
-    else:
-      source_id = sources_map[source_name]
-
-    address_delta = address - last_address
-    source_id_delta = source_id - last_source_id
-    line_delta = line - last_line
-    column_delta = column - last_column
-    mappings.append(encode_vlq(address_delta) + encode_vlq(source_id_delta) + encode_vlq(line_delta) + encode_vlq(column_delta))
-    last_address = address
-    last_source_id = source_id
-    last_line = line
-    last_column = column
-  return OrderedDict([('version', 3),
-                      ('names', []),
-                      ('sources', sources),
-                      ('sourcesContent', sources_content),
-                      ('mappings', ','.join(mappings))])
+    sources[source_id] = source_name
+    if collect_sources:
+      load_name = prefixes.load.resolve(file_name)
+      try:
+        with open(load_name, 'r') as infile:
+          source_content = infile.read()
+        sources_content.append(source_content)
+      except:
+        print('Failed to read source: %s' % load_name)
+        sources_content.append(None)
+  map['sourcesContent'] = sources_content
+
+def dwarf_to_json(wasm):
+  logging.debug('Read %s and convert to json' % wasm)
+
+  js = """const fs = require("fs"); const path = require("path");
+const __scriptdir = path.dirname(process.argv[1]);
+async function convertDwarfToJSON(input, enableXScopes = false) {
+    const dwarf_to_json_code = fs.readFileSync(__scriptdir + "/dwarf_to_json.wasm");
+    const {instance: {exports: { alloc_mem, free_mem, convert_dwarf, memory }}} =
+        await WebAssembly.instantiate(dwarf_to_json_code);
+    wasm = fs.readFileSync(input)
+    const wasmPtr = alloc_mem(wasm.byteLength);
+    new Uint8Array(memory.buffer, wasmPtr, wasm.byteLength).set(new Uint8Array(wasm));
+    const resultPtr = alloc_mem(12);
+    convert_dwarf(wasmPtr, wasm.byteLength, resultPtr, resultPtr + 4, enableXScopes);
+    free_mem(wasmPtr);
+    const resultView = new DataView(memory.buffer, resultPtr, 12);
+    const outputPtr = resultView.getUint32(0, true), outputLen = resultView.getUint32(4, true);
+    free_mem(resultPtr);
+    const output = Buffer.from(new Uint8Array(memory.buffer, outputPtr, outputLen)).toString("utf8");
+    free_mem(outputPtr);
+    return output;
+}
```

Comment:
Perhaps put this in a separate file?  dwarf_to_json.js?

We should probably include documentation for on where dwarf_to_json.wasm comes from and how to rebuild it?  I assume its a rust project built to wasm?    Its pretty cool to use wasm here but is there a reason to prefer it over using the rust native binary?   I guess we get the single binary thing? 

One alternative would be to ship the OS/arch-specific binary?  Could be tricky to build it for all platforms I guess?  But since speed is what we are after here perhaps native would be better?  I'm not sure.

This is cool either way :)

Reply:
> We should probably include documentation for on where dwarf_to_json.wasm comes from and how to rebuild it?

It is built from Rust into wasm.

>  Its pretty cool to use wasm here but is there a reason to prefer it over using the rust native binary?  I guess we get the single binary thing?

Correct. It is to avoid OS/arch-specific binaries.

Not sure how doable is to make Rust part of the emscripten build process. But it will be also beneficial to (re)write entire wasm-sourcemap.py to be more optimal. This is the first step to replace dependency on llvm-dwarfdump.